### PR TITLE
Feature release/add metrics to preview

### PR DESCRIPTION
**Pull request by yaojiefeng**
reopen #8456 against release 4.1
JIRA: https://issues.cask.co/browse/CDAP-9165
Build: http://builds.cask.co/browse/CDAP-RUT861-1

### CODE REVIEW

**Comment by tonybach** [comment]
Will also add UI commits to this PR, so the UI is not dependent on this being merged.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 1]
```diff
@@ -87,24 +93,64 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
...
+    if (this.runId) {
+      myLogsApi.getLogsMetadata({
+        namespace : this.namespaceId,
+        appId : this.appId,
+        programType : this.programType,
+        programId : this.programId,
+        runId : this.runId
+      }).$promise.then(
+        (res) => {
+          if (pollPromise && res.status === 'KILLED' || res.status==='COMPLETED' || res.status === 'FAILED' || res.status === 'STOPPED' || res.status === 'KILLED_BY_TIMER') {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: Looks simpler 
```
let previewStatus =  ['KILLED', 'COMPLETED', 'FAILED', 'STOPPED', 'KILLED_BY_TIMER'];
if (pollPromise && previewStatus.indexOf(res.status) !== -1) {
  ...
}
```

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 2]
```diff
@@ -87,24 +93,64 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
...
+            dataSrc.stopPoll(pollPromise.__pollId__);
+            pollPromise = null;
+          }
+        },
+        (err) => {
+          console.log('ERROR: ', err);
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: We don't want to surface this anywhere? Could this even error out?

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 3]
```diff
@@ -87,24 +93,64 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
...
+        });
+    } else if (this.previewId) {
+      myPreviewLogsApi.getLogsStatus({
+        namespace : this.namespaceId,
+        previewId : this.previewId
+      }).$promise.then(
+        (res) => {
+          if (pollPromise && res.status === 'KILLED' || res.status==='COMPLETED' || res.status === 'FAILED' || res.status === 'STOPPED' || res.status === 'KILLED_BY_TIMER') {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: Same here as well. We could share the `previewStatus` variable.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 4]
```diff
@@ -138,32 +184,54 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
 
   screenSize = LogViewerStore.getState().fullScreen;
 
-  if (!this.namespaceId || !this.appId || !this.programType || !this.programId || !this.runId) {
+  if (!(this.namespaceId && this.previewId) && !(this.namespaceId && this.appId && this.programType && this.programId && this.runId)) {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: I'm not exactly sure what is going on here. If I am not wrong are we checking if group of these values are present?

If yes can we put it in a variable such that it is simpler to read?

```
let validPreview = this.namespaceId && this.previewId;
let validProgram = this.namespaceId && this.appId && this.programType && this.programId && this.runId;

if (!validPreview && !validProgram) {
  ... 
 }
```

File: cdap-ui/app/directives/timeline/timeline.js [hunk 5]
```diff
@@ -507,7 +507,8 @@ angular.module(PKG.name + '.commons')
       appId: '@',
       programType: '@',
       programId: '@',
-      runId: '@'
+      runId: '@',
+      previewId: '@'
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js]: I thought we were going to create a copy of timeline directive to isolate changes to preview. Based on changes to timeline I don't know if we should add those changes to timeline directive being shared between different log viewers.

**Review by ajainarayanan — COMMENTED** [review]
Reviewed UI part of the PR. @tonybach

File: cdap-ui/app/directives/timeline/timeline-ctrl.js (hunk 2 above)
Comment (tonybach) [reply cdap-ui/app/directives/timeline/timeline-ctrl.js]: Not really sure what to do here. The purpose of this function is to check for the preview status, and if it's killed/stopped/completed, then stop the poll. So if we get an error while trying to get the status, maybe we can stop the poll anyways?

**Comment by tonybach** [comment]
Will also add commits for these JIRAs: [#9083](https://issues.cask.co/browse/CDAP-9083), [#9233](https://issues.cask.co/browse/CDAP-9233), [#9248](https://issues.cask.co/browse/CDAP-9248),  [#9252](https://issues.cask.co/browse/CDAP-9252).

Here are the screenshots showing the fix:
- #9083: Expanding an error/info log makes the text go underneath the previous log message.
screen shot 2017-04-05 at 12 12 35 pm

- #9233: Modals show a shadow when opened in studio
screen shot 2017-04-05 at 12 16 32 pm

- #9248: Run any batch/realtime preview/pipeline, you should see the total messages/errors/warnings correctly.

- #9252: Add a text saying that the preview pipeline realtime will run for 1 minute.
screen shot 2017-04-05 at 12 18 18 pm

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 6]
```diff
@@ -143,33 +148,32 @@ public void startPreview(PreviewRequest<?> previewRequest) throws Exception {
...
         setStatus(new PreviewStatus(PreviewStatus.Status.RUNNING, null, System.currentTimeMillis(), null));
+        // Only have timer if there is a timeout setting.
+        if (previewConfig.getTimeout() != null) {
+          timer = new Timer();
+          final int timeout =  previewConfig.getTimeout();
+          timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+              try {
+                LOG.info("Stopping the preview since it has reached running time: {} mins.", timeout);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java]: is it minutes ? the API does not say which time-unit the timeout is.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 7]
```diff
@@ -268,6 +271,7 @@ function LogViewerController ($scope, $window, LogViewerStore, myLogsApi, LOGVIE
       //otherwise, it does not have stack trace but has been selected
       vm.displayData[index].selected = !vm.displayData[index].selected;
     }
+    checkForScrollbar();
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js:274]: Do we need the same changes in log-viewer as well?

File: cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      getStatus();
+    }, (err) => {
+      // FIXME: We need to fix this. Right now this fails and we need to handle this more gracefully.
+      $scope.initialize();
+      console.log('ERROR: ', err);
+    });
+  };
+
+  LogViewerStore.subscribe(() => {
+    if(screenSize !== LogViewerStore.getState().fullScreen){
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js]: Seems like this is declared at the top of the file. Any specific reason why this is not attached to `this`. It took a while to see where this was declared.

File: cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      $scope.updateSliderHandle(LogViewerStore.getState().startTime);
+
+      //Check if the pinScrollPosition is less than the value of the query handle
+      this.pinScrollPosition = LogViewerStore.getState().scrollPosition;
+      if(typeof $scope.updatePin !== 'undefined'){
+        $scope.pinScrollingPosition = this.pinScrollPosition;
+        $scope.updatePin();
+      }
+
+      if($scope.searchResultTimes !== LogViewerStore.getState().searchResults){
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js]: We seem to be doing too many `LogViewerStore.getState()`. Can we have a `let state = LogViewerStore.getState()` and use state there after? Could avoid repeated function calls.

File: cdap-ui/app/directives/timeline-preview/timeline-preview.js [hunk 10]
```diff
@@ -0,0 +1,514 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline-preview/timeline-preview.js]: 2017

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 11]
```diff
@@ -77,7 +77,7 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
       names: ['system.app.log.error', 'system.app.log.warn', 'system.app.log.info', 'system.app.log.debug'],
       startTime : '',
       endTime : '',
-      resolution: '1m'
+      resolution: 'auto'
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: Was this verified from backend? Is it ok for UI to query metrics with a resolution set to `auto`?

**Review by ajainarayanan — COMMENTED** [review]
@tonybach Minor comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java (hunk 6 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java]: Yes, the timeout unit is minutes.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js (hunk 11 above)
Comment (tonybach) [reply cdap-ui/app/directives/timeline/timeline-ctrl.js]: Yes that's a fix recommended by the backend for this JIRA https://issues.cask.co/browse/CDAP-9248

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 7 above)
Comment (tonybach) [reply cdap-ui/app/directives/log-viewer/log-viewer.js:274]: Not sure what you mean, I added this change to both `log-viewer.js` and `log-viewer-preview.js`

Comment (ajainarayanan) [reply cdap-ui/app/directives/log-viewer/log-viewer.js:274]: Talked to @tonybach. This is a problem in the existing log viewer too (table headers won't align with table columns if scroll bar appears).

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewRunner.java [hunk 12]
```diff
@@ -18,6 +18,7 @@
 
 import co.cask.cdap.api.metrics.MetricTimeSeries;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewRunner.java]: unused import ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java [hunk 13]
```diff
@@ -63,9 +69,12 @@
 @Singleton
 @Path(Constants.Gateway.API_VERSION_3 + "/namespaces/{namespace-id}")
 public class PreviewHttpHandler extends AbstractLogHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewHttpHandler.class);
   private static final Gson GSON = new GsonBuilder()
     .registerTypeAdapter(BasicThrowable.class, new BasicThrowableCodec()).create();
   private static final Type STRING_LIST_MAP_TYPE = new TypeToken<Map<String, List<String>>>() { }.getType();
+  private static final String NAMESPACE_STRING = "namespace";
+  private static final String APP_STRING = "app";
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java]: i think i asked you to use the constants metrics tags in the previous PR.

**Review by gsps1 — COMMENTED** [review]
minor comments, fix and merge.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java (hunk 13 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java]: That constant has value "ns", but here user will pass tags as "namespace", so I cannot use that constant.

**Comment by ajainarayanan** [comment]
@tonybach Noticed two things in the behavior,
1. When we start a preview and keep the logs opened I was expecting to have it updated as soon as the preview starts to run. But had to close the log viewer and reopen it again. 
  From an implementation perspective I understand that we are passing in the context for the log viewer but wanted to check if it will be a big change to poll for new run preview id when the log viewer is opened. Would be a nice user experience
2. The metrics are surprisingly ahead of time, as in the count of the messages appear before the actual logs appear in the UI. See the screenshot for reference 

screen shot 2017-04-07 at 2 26 04 pm

The above screenshot shows 18 messages whereas in UI we still show less than 10 message. It could be that the UI polling happens every 10 seconds and thats the 10 seconds delay I am seeing when in UI. Could we check if reducing the polling interval fixes this?

If for the above mentioned points it looks awesome! 🎉 👍 LGTM on the code

**Comment by bdmogal** [comment]
Regarding @ajainarayanan's first comment above, immediately after clicking start preview in the run dropdown, the logs in the logviewer should be emptied.

Other than that, LGTM

**Comment by yaojiefeng** [comment]
@ajainarayanan @bdmogal  For the metrics part, the total messages displayed contains the log level DEBUG, which may make it larger than you see. Do we need the total messages updated with the log level you specify or keep it showing numbers of DEBUG, INFO, WARN, and ERROR

File: cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+  this.dataSrc = new MyCDAPDataSource($scope);
+  this.pinScrollPosition = 0;
+  this.screenSize = LogViewerStore.getState().fullScreen;
+  this.pollPromise = null;
+
+  this.apiSettings = {
+    metric : {
+      context: `namespace.${this.namespaceId}.app.${this.previewId}`,
+      names: ['system.app.log.error', 'system.app.log.warn', 'system.app.log.info', 'system.app.log.debug'],
```
Comment (bdmogal) [inline cdap-ui/app/directives/timeline-preview/timeline-ctrl-preview.js]: Why do we query only these 4? Why not trace?

**Pull request merged**